hugr/src/types/type_param.rs

```diff
@@ -396,3 +469,172 @@ pub enum TypeArgError {
+        check(USIZE_T, &TypeBound::Eq.into()).unwrap();
+        let seq_param = TypeParam::new_list(TypeBound::Eq);
+        check(USIZE_T, &seq_param).unwrap_err();
+        check_seq(&[USIZE_T], &TypeBound::Any.into()).unwrap_err();
+
+        // Into a list of type, we can fit a single row var
+        check(rowvar(0, TypeBound::Eq), &seq_param).unwrap();
+        // or a list of (types or row vars)
+        check(vec![], &seq_param).unwrap();
+        check_seq(&[rowvar(0, TypeBound::Eq)], &seq_param).unwrap();
+        check_seq(
+            &[rowvar(1, TypeBound::Any), USIZE_T, rowvar(0, TypeBound::Eq)],
+            &TypeParam::new_list(TypeBound::Any),
+        )
+        .unwrap();
+        // Next one fails because a list of Eq is required
+        check_seq(
+            &[rowvar(1, TypeBound::Any), USIZE_T, rowvar(0, TypeBound::Eq)],
+            &seq_param,
+        )
+        .unwrap_err();
+        // seq of seq of types is not allowed
+        check(
+            vec![USIZE_T.into(), vec![USIZE_T.into()].into()],
+            &seq_param,
+        )
+        .unwrap_err();
+
+        // Similar for nats (but no equivalent of fancy row vars)
+        check(5, &TypeParam::max_nat()).unwrap();
+        check_seq(&[5], &TypeParam::max_nat()).unwrap_err();
+        let list_of_nat = TypeParam::new_list(TypeParam::max_nat());
+        check(5, &list_of_nat).unwrap_err();
+        check_seq(&[5], &list_of_nat).unwrap();
+        check(TypeArg::new_var_use(0, list_of_nat.clone()), &list_of_nat).unwrap();
+        // But no equivalent of row vars - can't append a nat onto a list-in-a-var:
+        check(
+            vec![5.into(), TypeArg::new_var_use(0, list_of_nat.clone())],
+            &list_of_nat,
+        )
+        .unwrap_err();
+
+        // TypeParam::Tuples require a TypeArg::Seq of the same number of elems
+        let usize_and_ty = TypeParam::Tuple {
+            params: vec![TypeParam::max_nat(), TypeBound::Eq.into()],
+        };
+        check(vec![5.into(), USIZE_T.into()], &usize_and_ty).unwrap();
+        check(vec![USIZE_T.into(), 5.into()], &usize_and_ty).unwrap_err(); // Wrong way around
+        let two_types = TypeParam::Tuple {
+            params: vec![TypeBound::Any.into(), TypeBound::Any.into()],
+        };
+        check(TypeArg::new_var_use(0, two_types.clone()), &two_types).unwrap();
+        // not a Row Var which could have any number of elems
+        check(TypeArg::new_var_use(0, seq_param), &two_types).unwrap_err();
+    }
+
+    #[test]
+    fn type_arg_subst_row() {
+        let row_param = TypeParam::new_list(TypeBound::Copyable);
+        let row_arg = TypeArg::Sequence {
+            elems: vec![BOOL_T.into(), Type::UNIT.into()],
+        };
+        check_type_arg(&row_arg, &row_param).unwrap();
+
+        // Now say a row variable referring to *that* row was used
+        // to instantiate an outer "row parameter" (list of type).
```

Comment:
So I think I can do a non-unit-test of this - it requires defining a custom `TypeDef` parameterized by a list-of-type, then a Hugr FuncDef (parameterized by a list-of-type) returning an instance of that type-def (using the function's row-variable parameter). Would be significantly longer but I think it would work. Worth doing???